include/caffe/layer.hpp

```diff
@@ -490,7 +503,17 @@ template <typename Dtype>
 inline void Layer<Dtype>::Backward(const vector<Blob<Dtype>*>& top,
     const vector<bool>& propagate_down,
     const vector<Blob<Dtype>*>& bottom) {
-  switch (Caffe::mode()) {
+  Caffe::Brew mode;
+  if (layer_mode_ == LayerParameter_LayerMode_CPU) {
+    mode = Caffe::CPU;
+  } else if (layer_mode_ == LayerParameter_LayerMode_GPU) {
+    mode = Caffe::GPU;
+  } else if (layer_mode_ == LayerParameter_LayerMode_SOLVER) {
+    mode = Caffe::mode();
+  } else {
+    LOG(FATAL) << "Unknown layer mode.";
+  }
+  switch (mode) {
```

Comment:
This block of code is duplicated. Can we factor it out into a GetMode() method perhaps.